cpp/src/io/orc/dict_enc.cu

```diff
@@ -16,471 +16,261 @@
+    dim3(cudf::util::div_rounding_up_unsafe<unsigned int>(num_rowgroups, block_size),
+         static_cast<unsigned int>(num_str_cols));
 
-  dictinit_state_s* const s = &state_g;
-  // Index of the column in the `str_col_indexes` array
-  uint32_t const str_col_idx = blockIdx.x;
-  // Index of the column in the `orc_columns` array
-  auto const col_idx      = str_col_indexes[str_col_idx];
-  uint32_t group_id       = blockIdx.y;
-  auto const num_str_cols = str_col_indexes.size();
-  uint32_t nnz, start_row, dict_char_count;
-  int t = threadIdx.x;
+  rowgroup_char_counts_kernel<<<grid_size, block_size, 0, stream.value()>>>(
+    counts, orc_columns, rowgroup_bounds, str_col_indexes);
+}
 
-  if (t == 0) {
-    s->chunk             = chunks[group_id][str_col_idx];
-    s->chunk.leaf_column = &orc_columns[col_idx];
-    s->chunk.dict_data   = dict_data[str_col_idx].data() + rowgroup_bounds[group_id][col_idx].begin;
-    s->chunk.dict_index  = dict_index[str_col_idx].data();
-    s->chunk.start_row   = rowgroup_bounds[group_id][col_idx].begin;
-    s->chunk.num_rows    = rowgroup_bounds[group_id][col_idx].size();
-    s->dict              = tmp_indices[str_col_idx].data() + s->chunk.start_row;
-  }
-  for (uint32_t i = 0; i < sizeof(s->map) / sizeof(uint32_t); i += block_size) {
-    if (i + t < sizeof(s->map) / sizeof(uint32_t)) s->map.u32[i + t] = 0;
-  }
-  __syncthreads();
-  // First, take care of NULLs, and count how many strings we have (TODO: bypass this step when
-  // there are no nulls)
-  LoadNonNullIndices<block_size>(s, t, temp_storage.scan_storage);
-  // Sum the lengths of all the strings
-  if (t == 0) {
-    s->chunk.string_char_count = 0;
-    s->total_dupes             = 0;
-  }
-  nnz              = s->nnz;
-  auto t_dict_data = s->chunk.dict_data;
-  start_row        = s->chunk.start_row;
-  for (uint32_t i = 0; i < nnz; i += block_size) {
-    uint32_t ck_row = 0;
-    uint32_t hash   = 0;
-    uint32_t len    = 0;
-    if (i + t < nnz) {
-      ck_row                 = s->dict[i + t];
-      string_view string_val = s->chunk.leaf_column->element<string_view>(ck_row + start_row);
-      len                    = static_cast<uint32_t>(string_val.size_bytes());
-      hash                   = hash_string(string_val);
-    }
-    len = block_reduce(temp_storage.reduce_storage).Sum(len);
-    if (t == 0) s->chunk.string_char_count += len;
-    if (i + t < nnz) {
-      atomicAdd(&s->map.u32[hash >> 1], 1 << ((hash & 1) ? 16 : 0));
-      t_dict_data[i + t] = start_row + ck_row;
+template <int block_size>
+__global__ void __launch_bounds__(block_size)
+  initialize_dictionary_hash_maps_kernel(device_span<stripe_dictionary> dictionaries)
+{
+  auto const dict_map = dictionaries[blockIdx.x].map_slots;
+  auto const t        = threadIdx.x;
+  for (size_type i = 0; i < dict_map.size(); i += block_size) {
+    if (t + i < dict_map.size()) {
+      new (&dict_map[t + i].first) map_type::atomic_key_type{KEY_SENTINEL};
```

Comment:
Can we use just this?
```
auto& dict_map = ...;
dict_map[t + i] = {..., ...};
```

Reply:
I don't know why placement new is used here.
This is a question for @PointKernel, took this from his Parquet code :grin:

Reply:
I tried to assign, getting this error
` error: function "cuda::__4::atomic<_Tp, _Sco>::operator=(const cuda::__4::atomic<cudf::size_type, cuda::std::__4::__detail::thread_scope_device> &) [with _Tp=cudf::size_type, _Sco=cuda::std::__4::__detail::thread_scope_device]" (declared implicitly) cannot be referenced -- it is a deleted function`
I assume it's because of atomic types used here.